apiserver/facades/agent/secretsmanager/secrets.go

```diff
@@ -344,20 +344,93 @@ func (s *SecretsManagerAPI) GetSecretContentInfo(args params.GetSecretContentArg
 	return result, nil
 }
 
-func (s *SecretsManagerAPI) getSecretContent(arg params.GetSecretContentArg) (*secrets.ContentParams, error) {
-	uri, err := coresecrets.ParseURI(arg.URI)
+func (s *SecretsManagerAPI) getOwnerSecretMetadata(uri *coresecrets.URI, label string) (*coresecrets.SecretMetadata, error) {
+	if uri == nil {
+		var err error
+		uri, err = s.secretsBackend.GetURIBySecretLabel(label, s.authTag)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	md, err := s.secretsBackend.GetSecret(uri)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	if !s.canRead(uri, s.authTag) {
-		return nil, apiservererrors.ErrPerm
+	if s.authTag.String() != md.OwnerTag {
+		// Invisible for non owner users.
+		return nil, errors.NotFoundf("secret %s, label %q", uri, label)
+	}
+	return md, nil
+}
+
+func (s *SecretsManagerAPI) getSecretContent(arg params.GetSecretContentArg) (*secrets.ContentParams, error) {
+	// Only the owner can access secrets via the secret label (Note: the leader unit is not the owner of the application secrets).
+	// Non owners have to use secret URI or consumer label.
+
+	if arg.URI == "" && arg.Label == "" {
+		return nil, errors.NewNotValid(nil, "both uri and label are empty")
 	}
+
+	getSecretValue := func(uri *coresecrets.URI, revision int) (*secrets.ContentParams, error) {
+		if !s.canRead(uri, s.authTag) {
+			return nil, apiservererrors.ErrPerm
+		}
+
+		val, providerId, err := s.secretsBackend.GetSecretValue(uri, revision)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return &secrets.ContentParams{SecretValue: val, ProviderId: providerId}, nil
+	}
+
+	var uri *coresecrets.URI
+	var err error
+
+	if arg.URI != "" {
+		uri, err = coresecrets.ParseURI(arg.URI)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	// arg.Label is the secret label for owner.
+	md, err := s.getOwnerSecretMetadata(uri, arg.Label)
+	if err != nil && !errors.Is(err, errors.NotFound) {
+		return nil, errors.Trace(err)
+	}
+	if md != nil {
+		// Owner can access the content directly.
+		if arg.Update {
+			// We should not have cmd flag knowledge here but we have to check here because jujuc cannot know it.
+			return nil, errors.NewNotValid(nil, "secret owner cannot use --update")
+		}
```

Comment:
We need to check this in the hook context before making this api call. The hook context has pre-loaded all the metadata fore secrets owned by the caller. So we can look up the owned secret URI in the hook context and just pass the URI to the GetSecretValue facade call.

See contextfactory.go
```
	info, err := ctx.secretsClient.SecretMetadata(secrets.Filter{
		OwnerTag: &owner,
	})
```

The only time we will not have a uri passed in is if a secret consumer is making the api call. In this case, we do not even look up any owner metadata - we just search for consumer metadata.